pkg/dialer/dialer.go

```diff
@@ -29,51 +30,85 @@ type dialResult struct {
 	err error
 }
 
-// ContextDialer returns a GRPC net.Conn connected to the provided address
+var (
+	// ErrTimeout represents a timeout error while dialing
+	ErrTimeout = stderrs.New("timeout")
+)
+
+// ContextDialer returns a GRPC net.Conn connected to the provided address.
+// It tolerates ENOENT and keeps retrying if provided context has a deadline.
+// It does a "one shot" connect attempt if provided context doesn't have a deadline.
 func ContextDialer(ctx context.Context, address string) (net.Conn, error) {
-	if deadline, ok := ctx.Deadline(); ok {
-		return timeoutDialer(address, time.Until(deadline))
+	if _, ok := ctx.Deadline(); !ok {
+		return contextDialer(ctx, address, nil)
 	}
-	return timeoutDialer(address, 0)
+	return contextDialer(ctx, address, isNoent)
+}
+
+// ContextDialerFunc returns a GRPC net.Conn connected to the provided address,
+// but may tolerate certain dial errors and keep retrying until context canceled.
+// If tolerateErr is nil, it will tolerate no error and return what it got immediately while dialing.
+func ContextDialerFunc(ctx context.Context, address string, tolerateErr func(error) bool) (net.Conn, error) {
+	return contextDialer(ctx, address, tolerateErr)
 }
 
 // Dialer returns a GRPC net.Conn connected to the provided address
 // Deprecated: use ContextDialer and grpc.WithContextDialer.
 var Dialer = timeoutDialer
 
-func timeoutDialer(address string, timeout time.Duration) (net.Conn, error) {
-	var (
-		stopC = make(chan struct{})
-		synC  = make(chan *dialResult)
-	)
+// timeoutDialer connects to the provided address with a timeout.
+func timeoutDialer(address string, timeout time.Duration) (conn net.Conn, err error) {
```

Comment:
Dialer/timeoutDialer was deprecated in favor of forcing the caller to set context. In the containerd/containerd repo we only use dialer.Dialer a few times.  Perhaps we should stop using Dialer in those and just go ahead and recode to use proper context?

Reply:
However.. it's exported and located under `pkg`, no one knows if there are any 3rd party depends on it. If any member from containerd determines to completely remove the deprecated `Dialer`, I am glad to help remove any existing usage in the project scope.